plugins/validation/src/main/nextflow/validation/SchemaValidator.groovy

```diff
@@ -159,14 +324,72 @@ class SchemaValidator {
             else {
                 def param = exJSON['pointerToViolation'] - ~/^#\//
                 def param_val = paramsJSON[param].toString()
-                errors << "* --${param}: ${exJSON['message']} (${param_val})".toString()
+                if (enums.containsKey(param)) {
+                    def error_msg = "* --${param}: '${param_val}' is not a valid choice (Available choices"
+                    if (enums[param].size() > limit) {
+                        errors << "${error_msg} (${limit} of ${enums[param].size()}): ${enums[param][0..limit-1].join(', ')}, ... )".toString()
+                    } else {
+                        errors << "${error_msg}: ${enums[param].join(', ')})".toString()
+                    }
+                } else {
+                    errors << "* --${param}: ${exJSON['message']} (${param_val})".toString()
+                }
             }
         }
         for (ex in causingExceptions) {
-            collectErrors(ex, paramsJSON)
+            collectErrors(ex, paramsJSON, enums)
         }
     }
 
+    //
+    // Remove an element from a JSONArray
+    //
+    private static JSONArray removeElement(json_array, element) {
+        def list = []
+        int len = json_array.length()
+        for (int i=0;i<len;i++){
+            list.add(json_array.get(i).toString())
+        }
+        list.remove(element)
+        JSONArray jsArray = new JSONArray(list)
+        return jsArray
+    }
```

Comment:
@ewels I copied the whole functions here even though they are not used by the plugin. But I don't know if we may want to still have them and maybe call them. For example this one.
Maybe I am not fully understanding how the schema validation works. Will we keep the NfcoreShcema.groovy file in nf-core to still use these functions? For example paramsHelp() that you mentioned in your comment. I understood we were getting rid of the complete file.

Reply:
Not `paramsHelp()` as that's used in workflows (see my [comment above](https://github.com/nextflow-io/nf-validation/pull/5#issuecomment-1145880306)):

> For example, paramsHelp() is used in [lib/WorkflowMain.groovy](https://github.com/nf-core/tools/blob/171127bd850040e76febd0945e6980b7afcaad69/nf_core/pipeline-template/lib/WorkflowMain.groovy#L28): `help_string += NfcoreSchema.paramsHelp(workflow, params, command)`

The others are nearly all helper functions used in [`NfcoreSchema.groovy`](https://github.com/nf-core/tools/blob/171127bd850040e76febd0945e6980b7afcaad69/nf_core/pipeline-template/lib/NfcoreSchema.groovy)  in the template, so assuming that the code is comparable between the two then they can't be removed either.

Reply:
They all seem to be used, mostly in this file in this PR:

- `paramsHelp()`- pipeline ` lib/WorkflowMain.groovy`
- `paramsMaxChars()`- L170, L292
- `paramsLoad()`- L169, L242
- `paramsRead()`- L422
- `paramsMaxChars()`- L170, L292
- `paramsSummaryLog()`- pipeline `lib/WorkflowMain.groovy`
- `paramsSummaryMap()`- L291
- `removeIgnoredParams()`- [L144](https://github.com/nf-core/tools/blob/b1ce47148a05965e0604534cece3359d408af357/nf_core/pipeline-template/lib/NfcoreSchema.groovy#L144) of `lib/NfcoreSchema.groovy` but looks like it hasn't been carried over into this PR.

Reply:
If I understand well, `paramsHelp()` will only be used in nf-core pipelines, but not within the plugin, so can be removed from this PR, right?

Reply:
And yeah, maybe we should also rename the plugin to be `nf-schema` instead of `nf-validation`...

Reply:
OK I get it now, thanks! Then I will need to test this extra functionality and see if I need to adapt something else here